crates/common/types/fork_id.rs

```diff
@@ -48,12 +49,88 @@ impl ForkId {
             fork_next,
         }
     }
+
+    // See https://eips.ethereum.org/EIPS/eip-2124#validation-rules.
+    pub fn is_valid(
+        &self,
+        incoming: Self,
+        latest_block_number: u64,
+        head_timestamp: u64,
+        chain_config: ChainConfig,
+        genesis_hash: BlockHash,
+    ) -> bool {
+        let (block_number_based_forks, timestamp_based_forks) = chain_config.gather_forks();
+        // decide if our head is block or timestamp based.
+        let mut head = head_timestamp;
+        if let Some(last_block_number_based_fork) = block_number_based_forks.last() {
+            if *last_block_number_based_fork > latest_block_number {
+                head = latest_block_number;
+            }
+        }
+        if incoming.fork_hash == self.fork_hash {
+            // validation rule #1
+            if incoming.fork_next == 0 {
+                return true;
+            }
+            if incoming.fork_next <= head {
+                debug!("Future fork already passed locally.");
+                return false;
+            }
+            return true;
+        }
+
+        let forks = [block_number_based_forks, timestamp_based_forks].concat();
+        let valid_combinations = get_all_fork_id_combinations(forks, genesis_hash);
+
+        let mut is_subset = true;
+
+        for (fork_hash, fork_next) in valid_combinations {
+            if is_subset {
+                // is a subset of the local past forks (rule #2)
+                if incoming.fork_hash == fork_hash && incoming.fork_next == fork_next {
+                    return true;
+                }
+            } else {
+                // is a superset of the local past forks (rule #3)
+                if incoming.fork_hash == fork_hash {
+                    return true;
+                }
+            }
+            if fork_hash == self.fork_hash {
+                // from this point, is a superset of the local past forks
```

Comment:
I'd reword this a bit

```suggestion
                // From this point on, if we have a match, it means the remote fork hash
                // is a superset of our local past forks.
```